File: clients/android/feature/vault/src/main/kotlin/com/clipshare/feature/vault/VaultFeatureModels.kt
package com.clipshare.feature.vault

import com.clipshare.core.vault.SyncPolicy
import com.clipshare.core.vault.SyncPolicySetting
import com.clipshare.core.vault.VaultContentType
import java.time.Instant

enum class VaultSection {
    INBOX,
    ALL,
    FOLDERS,
    RECENT,
    SYNCED,
    LOCAL_ONLY,
    TRASH,
    DEVICES,
}

enum class VaultEntityKind {
    FOLDER,
    ITEM,
}

data class VaultEntitySelection(val kind: VaultEntityKind, val id: String)

data class VaultFolderView(
    val id: String,
    val parentId: String?,
    val name: String,
    val description: String?,
    val sortOrder: Long,
    val policy: SyncPolicySetting,
    val effectivePolicy: SyncPolicy,
    val deleted: Boolean,
)

data class VaultItemView(
    val id: String,
    val folderId: String,
    val contentType: VaultContentType,
    val title: String,
    val text: String?,
    val fileName: String?,
    val sizeBytes: Long,
    val policy: SyncPolicySetting,
    val effectivePolicy: SyncPolicy,
    val createdAt: Instant,
    val updatedAt: Instant,
    val deleted: Boolean,
)

data class VaultDeviceView(val id: String, val displayName: String, val revoked: Boolean)

data class VaultSnapshot(
    val folders: List<VaultFolderView> = emptyList(),
    val items: List<VaultItemView> = emptyList(),
    val devices: List<VaultDeviceView> = emptyList(),
    val usedSlowSearch: Boolean = false,
    val notice: String? = null,
)

data class VaultFeatureState(
    val section: VaultSection = VaultSection.INBOX,
    val snapshot: VaultSnapshot = VaultSnapshot(),
    val selection: Set<VaultEntitySelection> = emptySet(),
    val busy: Boolean = false,
    val status: String? = null,
    val failure: Boolean = false,
    val draftTitle: String = "",
    val draftText: String = "",
    val draftContentType: VaultContentType = VaultContentType.TEXT,
    val sensitiveGeneration: Long = 0,
)

data class CreateVaultFolderCommand(val name: String, val parentId: String?, val description: String? = null)

data class RenameVaultFolderCommand(val folderId: String, val name: String, val description: String? = null)

data class SaveVaultTextCommand(
    val folderId: String,
    val title: String,
    val text: String,
    val contentType: VaultContentType,
)

data class ImportVaultFileCommand(val folderId: String, val title: String, val uri: String)

data class ExportVaultFileCommand(val itemId: String, val uri: String)

data class UpdateVaultTextCommand(val itemId: String, val title: String, val text: String)

data class VaultPolicyCommand(
    val policy: SyncPolicy,
    val selectedDeviceIds: Set<String> = emptySet(),
    val clearDescendantOverrides: Boolean = false,
)

File: clients/windows/src/ClipShare.Windows.Features.Vault/VaultFeatureModels.cs
namespace ClipShare.Windows.Features.Vault;

using ClipShare.Windows.Vault;

public enum VaultSection
{
    Inbox,
    All,
    Folders,
    Recent,
    Synced,
    LocalOnly,
    Trash,
    Devices,
}

public enum VaultEntityKind
{
    Folder,
    Item,
}

public readonly record struct VaultEntitySelection(VaultEntityKind Kind, string Id);

public sealed record VaultFolderView(
    string Id,
    string? ParentId,
    string Name,
    string? Description,
    long SortOrder,
    SyncPolicySetting Policy,
    SyncPolicy EffectivePolicy,
    bool Deleted);

public sealed record VaultItemView(
    string Id,
    string FolderId,
    VaultContentType ContentType,
    string Title,
    string? Text,
    string? FileName,
    long SizeBytes,
    SyncPolicySetting Policy,
    SyncPolicy EffectivePolicy,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt,
    bool Deleted);

public sealed record VaultDeviceView(string Id, string DisplayName, bool Revoked);

public sealed record VaultSnapshot(
    IReadOnlyList<VaultFolderView> Folders,
    IReadOnlyList<VaultItemView> Items,
    IReadOnlyList<VaultDeviceView> Devices,
    bool UsedSlowSearch,
    string? Notice);

public sealed record VaultFeatureState(
    VaultSection Section,
    VaultSnapshot Snapshot,
    IReadOnlySet<VaultEntitySelection> Selection,
    bool IsBusy,
    string? Status,
    bool IsFailure)
{
    public static VaultFeatureState Empty { get; } = new(
        VaultSection.Inbox,
        new VaultSnapshot([], [], [], false, null),
        new HashSet<VaultEntitySelection>(),
        false,
        null,
        false);
}

public sealed record CreateVaultFolderCommand(
    string Name,
    string? ParentId,
    string? Description = null);

public sealed record RenameVaultFolderCommand(string FolderId, string Name, string? Description = null);

public sealed record SaveVaultTextCommand(
    string FolderId,
    string Title,
    string Text,
    VaultContentType ContentType);

public sealed record ImportVaultFileCommand(
    string FolderId,
    string Title,
    string UserSelectedPath);

public sealed record UpdateVaultTextCommand(string ItemId, string Title, string Text);

public sealed record VaultPolicyCommand(
    SyncPolicy Policy,
    IReadOnlySet<string> SelectedDeviceIds,
    bool ClearDescendantOverrides);
